src/core/load-dropin.c

```diff
@@ -19,51 +19,245 @@
+        Unit *o;
+
+        /* Let's find out if we're involved in the deferred dep of
+         * another unit and the later is waiting for us to solve the
+         * dropin dep, IOW if we're the last loaded deferred dep. */
+        SET_FOREACH(o, u->deferred_dropin_units, i)
+                unit_load_solve_dropin_dependencies(o);
+
+        /* We shouldn't be called again for this unit as it is going
+         * to be in loaded state, so 'deferred_dropin_units' set
+         * shouldn't be needed anymore. */
+        u->deferred_dropin_units = set_free(u->deferred_dropin_units);
+}
+
 static int add_dependency_consumer(
                 UnitDependency dependency,
                 const char *entry,
-                const char* filepath,
+                const char *filepath,
                 void *arg) {
-        Unit *u = arg;
+        _cleanup_free_ DropinSymlink *ds = NULL;
+        DependencyConsumerData *data = arg;
+        Unit *u, *other;
         int r;
 
-        assert(u);
+        assert(data);
+
+        u = data->unit;
+
+        /* Here's the deal: the dependency unit needs to be loaded so we can
+         * deal with aliases (if any) properly. But as we're already in the
+         * loading path this cannot be achieved here (if it's needed)...
+         * Therefore we defer the handling of the dropins during their own
+         * unit loading. */
+        other = manager_get_unit(u->manager, entry);
+        if (!other) {
+                /* Force 'filepath' to NULL because the symlink is not supposed
+                 * to define any aliases and a symlink to /dev/null indicates
+                 * a dropin mask, not that the unit should be masked. */
+                r = manager_load_unit_prepare(u->manager, entry, NULL, NULL, &other);
+                if (r < 0)
+                        return r;
+        }
+
+        r = set_ensure_allocated(&u->deferred_dropin_dependencies, &dropin_hash_ops);
+        if (r < 0)
+                return r;
 
-        r = unit_add_dependency_by_name(u, dependency, entry, filepath, true);
+        r = set_ensure_allocated(&other->deferred_dropin_units, NULL);
         if (r < 0)
-                log_error_errno(r, "Cannot add dependency %s to %s, ignoring: %m", entry, u->id);
+                return r;
+
+        ds = new0(DropinSymlink, 1);
+        if (!ds)
+                return -ENOMEM;
 
+        ds->unit = other;
+        ds->depth = data->depth;
+        ds->dependency = dependency;
+        ds->is_mask = symlink_to_devnull(filepath);
+
+        r = set_put(u->deferred_dropin_dependencies, ds);
+        if (r <= 0)
+                /* r == 0 means we already registered the same dependency
+                 * for 'u' but with a higher priority (lower depth). */
+                return r;
+
+        r = set_put(other->deferred_dropin_units, u);
+        if (r < 0) {
+                set_remove(u->deferred_dropin_dependencies, ds);
+                return r;
+        }
```

Comment:
I am not sure I grok why we need two sets here, one index by the unit+dep, and one by the unit...

I mean, wouldn't it suffice to have the unit+dep one? why the other one?